examples/shipping/booking/instrumenting.go

```diff
@@ -0,0 +1,114 @@
+package booking
+
+import (
+	"time"
+
+	stdprometheus "github.com/prometheus/client_golang/prometheus"
+
+	"github.com/go-kit/kit/metrics"
+	kitprometheus "github.com/go-kit/kit/metrics/prometheus"
+
+	"github.com/go-kit/kit/examples/shipping/cargo"
+	"github.com/go-kit/kit/examples/shipping/location"
+)
+
+type instrumentingService struct {
+	requestCount   metrics.Counter
+	requestLatency metrics.TimeHistogram
+	Service
+}
+
+// NewInstrumentingService returns an instance of an instrumenting Service.
+func NewInstrumentingService(s Service) Service {
+	fieldKeys := []string{"method"}
+
+	requestCount := kitprometheus.NewCounter(stdprometheus.CounterOpts{
+		Namespace: "api",
+		Subsystem: "booking_service",
+		Name:      "request_count",
+		Help:      "Number of requests received.",
+	}, fieldKeys)
+
+	requestLatency := metrics.NewTimeHistogram(time.Microsecond, kitprometheus.NewSummary(stdprometheus.SummaryOpts{
+		Namespace: "api",
+		Subsystem: "booking_service",
+		Name:      "request_latency_microseconds",
+		Help:      "Total duration of requests in microseconds.",
+	}, fieldKeys))
+
+	return &instrumentingService{
+		requestCount:   requestCount,
+		requestLatency: requestLatency,
+		Service:        s,
+	}
+}
```

Comment:
I actually started off like that, but with multiple services, each with
multiple metrics, I felt my main.go got a little out of hand. I don't feel
too strongly about it though and I wouldn't have any problems with changing
it back. Didn't think of the registration management though, that's a good
point.

Den fre 27 maj 2016 16:59Peter Bourgon <notifications@github.com> skrev:

> In examples/shipping/booking/instrumenting.go
> <https://github.com/go-kit/kit/pull/278#discussion_r64917949>:
>
> > +		Help:      "Number of requests received.",
> > +	}, fieldKeys)
> > +
> > +	requestLatency := metrics.NewTimeHistogram(time.Microsecond, kitprometheus.NewSummary(stdprometheus.SummaryOpts{
> > +		Namespace: "api",
> > +		Subsystem: "booking_service",
> > +		Name:      "request_latency_microseconds",
> > +		Help:      "Total duration of requests in microseconds.",
> > +	}, fieldKeys))
> > +
> > +	return &instrumentingService{
> > +		requestCount:   requestCount,
> > +		requestLatency: requestLatency,
> > +		Service:        s,
> > +	}
> > +}
>
> Hmm. I tend to take the metrics as dependencies, rather than constructing
> them in the middleware, as you sometimes need to manage the registration
> (via the Prometheus HTTP handler) in unique ways. But I reckon that's a
> matter of opinion.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/go-kit/kit/pull/278/files/bab34e99795d04acc35b8da9f41037b787fd4e34#r64917949>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AIAgUL7g7PhiAkBZZr0jyp2jhebGaO4Pks5qFwbWgaJpZM4IoVaE>
> .
>